packages/lodestar/src/api/impl/beacon/state/utils.ts

```diff
@@ -32,21 +41,24 @@ export async function resolveStateId(
     return (await db.stateCache.get(forkChoice.getJustifiedCheckpoint().root)) ?? null;
   }
   if (stateId.startsWith("0x")) {
-    //TODO: support getting finalized states by root as well
-    return (await db.stateCache.get(fromHexString(stateId))) ?? null;
+    const stateRoot = fromHexString(stateId);
+    const cachedStateCtx = await db.stateCache.get(stateRoot);
+    if (cachedStateCtx) return cachedStateCtx;
+    const finalizedState = await db.stateArchive.getByRoot(stateRoot);
+    return stateToApiStateContext(config, finalizedState);
   }
   //block id must be slot
   const slot = parseInt(stateId, 10);
   if (isNaN(slot) && isNaN(slot - 0)) {
-    throw new Error("Invalid block id");
+    throw new Error("Invalid state id");
   }
-  //todo: resolve archive slot -> state
   const blockSummary = forkChoice.getCanonicalBlockSummaryAtSlot(slot);
   if (blockSummary) {
     return (await db.stateCache.get(blockSummary.stateRoot)) ?? null;
+  } else {
+    const finalizedState = await db.stateArchive.get(slot);
+    return stateToApiStateContext(config, finalizedState);
```

Comment:
we should probably split this method into bunch of utilities functions. I didn't expect it to grow that much tbh haha